src/Swarm/TUI/Model.hs

```diff
@@ -298,6 +307,61 @@ getCurrentItemText history = replItemText <$> Seq.lookup (history ^. replIndex)
 replIndexIsAtInput :: REPLHistory -> Bool
 replIndexIsAtInput repl = repl ^. replIndex == replLength repl
 
+------------------------------------------------------------
+-- Repl Prompt
+------------------------------------------------------------
+
+-- | This data type represent what is prompted to the player
+--   and how the REPL show interpret the user input.
+data REPLPrompt
+  = -- | Interpret the given text as a regular command
+    CmdPrompt Text
+  | -- | Interpret the given text as "search this text in history"
+    SearchPrompt Text REPLHistory
+
+-- | Get the last Entry in REPLHistory matching the given text
+lastEntry :: Text -> REPLHistory -> Maybe Text
+lastEntry t h = case filter (liftA2 (&&) matchesText isREPLEntry) $ toList $ h ^. replSeq of
+  [] -> Nothing
+  rh : _ -> Just $ replItemText rh
+ where
+  matchesText histItem = t `T.isInfixOf` replItemText histItem
+
+-- | The default REPL prompt.
+defaultPrompt :: REPLPrompt
+defaultPrompt = CmdPrompt ""
+
+--SearchPrompt "hea" (newREPLHistory [REPLEntry "def move", REPLEntry "head move"])
+
+-- | Lens for accesing the text of the prompt.
+promptTextL :: Lens' REPLPrompt Text
+promptTextL = lens g s
+ where
+  -- Notice that the prompt ADT must have a Text field in every constructor (representing what the user writes).
+  -- This should be force in the ADT itself... right know this here
+  -- The compiler will complain about "Non complete patterns" on this two function.
+  g :: REPLPrompt -> Text
+  g (CmdPrompt t) = t
+  g (SearchPrompt t _) = t
+
+  s :: REPLPrompt -> Text -> REPLPrompt
+  s (CmdPrompt _) t = CmdPrompt t
+  s (SearchPrompt _ h) t = SearchPrompt t h
+
+-- | Turn the repl prompt into a decorator for the form
+replPromptAsWidget :: REPLPrompt -> Widget Name
+replPromptAsWidget (CmdPrompt _) = txt "> "
+replPromptAsWidget (SearchPrompt t rh) =
+  case lastEntry t rh of
+    Nothing -> txt $ "[find command " <> t <> " nothing found] "
+    Just lastentry
+      | T.null t -> txt $ "[find command " <> t <> "] "
+      | otherwise -> txt $ "[find command " <> t <> ": \"" <> lastentry <> "\" found] "
```

Comment:
These prompts seem a bit too verbose.  I think we could get rid of the word `command` and also the word `found` at the end, and it would save some horizontal space and still make perfect sense.